activation/activation.go

```diff
@@ -496,133 +496,176 @@ func (b *Builder) BuildNIPostChallenge(ctx context.Context, nodeID types.NodeID)
 		return nil, ctx.Err()
 	case <-b.syncer.RegisterForATXSynced():
 	}
-	current := b.layerClock.CurrentLayer().GetEpoch()
 
-	challenge, err := nipost.Challenge(b.localDB, nodeID)
-	switch {
-	case errors.Is(err, sql.ErrNotFound):
-		// build new challenge
-		logger.Info("building new NiPOST challenge", zap.Uint32("current_epoch", current.Uint32()))
-	case err != nil:
-		logger.Info("failed to load NiPoST challenge from local state", zap.Error(err))
-		return nil, fmt.Errorf("get nipost challenge: %w", err)
-	case challenge.PublishEpoch < current:
-		logger.Info(
-			"existing NiPoST challenge is stale, resetting state",
-			zap.Uint32("current_epoch", current.Uint32()),
-			zap.Uint32("publish_epoch", challenge.PublishEpoch.Uint32()),
-		)
-		// Reset the state to idle because we won't be building POST until we get a new PoET proof
-		// (typically more than epoch time from now).
-		b.postStates.Set(nodeID, types.PostStateIdle)
-		if err := b.nipostBuilder.ResetState(nodeID); err != nil {
-			return nil, fmt.Errorf("reset nipost builder state: %w", err)
-		}
-		if err := nipost.RemoveChallenge(b.localDB, nodeID); err != nil {
-			return nil, fmt.Errorf("remove stale nipost challenge: %w", err)
-		}
-	default:
-		// challenge is fresh
-		logger.Info("loaded NiPoST challenge from local state",
-			zap.Uint32("current_epoch", current.Uint32()),
-			zap.Uint32("publish_epoch", challenge.PublishEpoch.Uint32()),
-		)
-		return challenge, nil
+	currentEpochId := b.layerClock.CurrentLayer().GetEpoch()
+
+	// try to get exising challenge
+	existingChallenge, err := b.getExistingChallenge(logger, currentEpochId, nodeID)
+	if err != nil && !errors.Is(err, sql.ErrNotFound) {
+		return nil, err
 	}
 
+	if existingChallenge != nil {
+		return existingChallenge, nil
+	}
```

Comment:
I think this would be easier if the `getExistingChallenge` method returned `nil, nil` in the case where the challenge was not in the database. Here we don't need to understand the sql errors. Consider:

```suggestion
	existingChallenge, err := b.getExistingChallenge(logger, currentEpochId, nodeID)
	if err != nil {
		return nil, fmt.Errorf("getting existing nipost challenge: %w", err)
	}

	if existingChallenge != nil {
		return existingChallenge, nil
	}
```